### PR TITLE
Refactor input to use shrev

### PR DESCRIPTION
**Pull request by Xaeroxe**
Problem with the current InputHandler: It queries for if a key was pressed or release `ThisFrame` which isn't very helpful to systems and state functions that don't trigger every frame.  Solution: Integrate shrev with the InputHandler so users can acquire and use `ReaderID`s with an `EventHandler<InputEvent>` specs resource.  More enhancements to the input system are on the way, but I just wanted to share this much since it is ready.

Notice: This PR also sneaks in a few other small improvements, such as Scancode buttons, and some new `new()` constructors.

cc @Rhuagh @jojolepro

### CODE REVIEW

**Comment by torkleyy** [comment]
I'm in favor of making `InputHandler` optional, as @Rhuagh suggested (together with allowing custom bindings). Not sure if that's what this PR is doing, I'll have to review this tomorrow.

**Comment by Rhuagh** [comment]
I'll have a PR for this PR soon, that does that.

**Comment by Xaeroxe** [comment]
Thanks for your PR @Rhuagh, I also had a few improvements of my own on top of your ideas.  This is ready for review now.

**Comment by jojolepro** [comment]
Code looks fine to me. I asked questions on gitter, and I'll approve when I have the answers ;)

File: amethyst_input/src/bindings.rs [hunk 1]
```diff
@@ -1,47 +1,62 @@
...
-        self.axes.keys().cloned().collect::<Vec<String>>()
+    pub fn axes(&self) -> Vec<AX> {
+        self.axes.keys().cloned().collect::<Vec<AX>>()
     }
 
     /// Add a button to an action.
     ///
     /// This will insert a new binding between this action and the button.
-    pub fn insert_action_binding<T: AsRef<str>>(&mut self, id: T, binding: Button) {
+    pub fn insert_action_binding(&mut self, id: AC, binding: Button) {
```
[the remark this answers is not in the document]
Comment (Rhuagh) [reply amethyst_input/src/bindings.rs]: I considered how that could be done, couldn't figure it out in the time I had, but our almighty omni figured it out :)

I don't foresee this method being called very often, but being less restrictive is always good.

File: src/ecs/input/system.rs [hunk 2]
```diff
@@ -41,13 +41,13 @@ where
 
     fn run(&mut self, (input, mut handler, mut output): Self::SystemData) {
         match input.read(&mut self.reader) {
-            Ok(data) => for d in data {
+            Ok(EventReadData::Data(data)) => for d in data {
                 if let &Event::WindowEvent { ref event, .. } = d {
                     handler.send_event(event, &mut *output);
                 }
             },
```
Comment (Rhuagh) [inline src/ecs/input/system.rs]: Might want a variant for EventReadData::Overflow that do the same things as above, and log a warning ?

**Comment by Xaeroxe** [comment]
@Aceeri I'd like you to review the InputEvent enum specifically and tell me if you think that gives you enough input data to make the UI framework.

**Comment by Rhuagh** [comment]
Controllers and vr headsets require support in winit first, but yes.

File: amethyst_input/src/bindings.rs [hunk 3]
```diff
@@ -1,47 +1,65 @@
...
 use smallvec::SmallVec;
 
 use super::{Axis, Button};
 
 /// Used for saving and loading input settings.
 #[derive(Default, Serialize, Deserialize, Clone)]
-pub struct Bindings {
-    pub(super) axes: HashMap<String, Axis>,
-    pub(super) actions: HashMap<String, SmallVec<[Button; 4]>>,
+pub struct Bindings<AX, AC> where AX: Hash + Eq, AC: Hash + Eq {
```
Comment (torkleyy) [inline amethyst_input/src/bindings.rs]: formatting?

**Comment by Xaeroxe** [comment]
@Rhuagh winit has stated that is out of scope: https://github.com/tomaka/winit/issues/119 so we'll need another crate that can gather input for that.

**Comment by torkleyy** [comment]
Looks great!

bors r+

**Pull request merged**